pkg/tbtc/node.go

```diff
@@ -205,3 +265,203 @@ func (n *node) joinDKGIfEligible(seed *big.Int, startBlockNumber uint64) {
+				dkg.ProtocolBlocks() +
+				delayBlocks
+		}
+
+		// Exclude all members controlled by the operators that were not
+		// qualified for the current attempt.
+		excludedMembers := make([]group.MemberIndex, 0)
+		attemptSkipped := false
+		for i, operator := range drl.selectedOperators {
+			if !qualifiedOperatorsSet[operator] {
+				memberIndex := group.MemberIndex(i + 1)
+				excludedMembers = append(excludedMembers, memberIndex)
+
+				// If the given member was not qualified for the given attempt,
+				// mark this attempt as skipped in order to skip the execution
+				// and set up the next attempt properly.
+				if memberIndex == drl.memberIndex {
+					attemptSkipped = true
+				}
+			}
+		}
+
+		var result *dkg.Result
+		var attemptErr error
+
+		if !attemptSkipped {
+			result, attemptErr = dkgAttemptFn(&dkgAttemptParams{
+				index:           drl.attemptCounter,
+				startBlock:      drl.attemptStartBlock,
+				excludedMembers: excludedMembers,
+			})
+			if attemptErr != nil {
+				var imErr *dkg.InactiveMembersError
+				if errors.As(attemptErr, &imErr) {
+					for _, memberIndex := range imErr.InactiveMembersIndexes {
+						operator := drl.selectedOperators[memberIndex-1]
+						drl.inactiveOperatorsSet[operator] = true
+					}
+				}
+			}
+		}
+
+		if attemptSkipped || attemptErr != nil {
+			var err error
+			qualifiedOperatorsSet, err = drl.qualifiedOperatorsSet()
+			if err != nil {
+				return nil, fmt.Errorf(
+					"cannot get qualified operators for attempt [%v]: [%w]",
+					drl.attemptCounter+1,
+					err,
+				)
+			}
+
+			continue
+		}
+
+		return result, nil
+	}
+}
+
+// qualifiedOperatorsSet returns a set of operators qualified to participate
+// in the given DKG attempt.
+func (drl *dkgRetryLoop) qualifiedOperatorsSet() (map[chain.Address]bool, error) {
+	// If this is one of the first attempts and random retries were not started
+	// yet, check if there are known inactive operators. If the group quorum
+	// can be maintained, just exclude the members controlled by the inactive
+	// operators from the qualified set.
+	if drl.attemptCounter <= 5 &&
+		drl.randomRetryCounter == 0 &&
+		len(drl.inactiveOperatorsSet) > 0 {
+		qualifiedOperators := make(chain.Addresses, 0)
+		for _, operator := range drl.selectedOperators {
+			if !drl.inactiveOperatorsSet[operator] {
+				qualifiedOperators = append(qualifiedOperators, operator)
+			}
+		}
+
+		if len(qualifiedOperators) >= drl.chainConfig.GroupQuorum {
+			return qualifiedOperators.Set(), nil
+		}
+	}
+
+	// In any other case, try to make a random retry.
+	qualifiedOperators, err := retry.EvaluateRetryParticipantsForKeyGeneration(
+		drl.selectedOperators,
+		drl.randomRetrySeed,
+		drl.randomRetryCounter,
+		uint(drl.chainConfig.GroupQuorum),
+	)
+	if err != nil {
+		return nil, fmt.Errorf(
+			"random operator selection failed: [%w]",
+			err,
+		)
+	}
+
+	drl.randomRetryCounter++
+	return chain.Addresses(qualifiedOperators).Set(), nil
```

Comment:
if we're going to immediately turn the list to a set, I can rewrite the retry API to return that directly and reduce complexity; I'm going out of the way to return a list at the moment

Reply:
See https://github.com/keep-network/keep-core/pull/3169#discussion_r953666514

Reply:
I read through the referenced comment, but I think this is potentially separate!

As I understand it, right now we have a need to represent the information as both a list of seats and a set of operators. The set of operators can always be inferred from the list of seats, but not the other way around and so the set of operators is *lossy*.

That said, are there other places where we intend to call `EvaluateRetryParticipantsForKeyGeneration`? Because here, we immediately turn the list of seats into a set of operators via `.Set()`. If we *always* plan on doing this (or are okay with some refactoring in the case where we thought we would always do this, but then later need the list), I can make `EvaluateRetryParticipantsForKeyGeneration` return a `Set` directly, and it would simplify a lot of that code, and the calling code as well.

Reply:
Honestly, I'd leave it as is. The current version is similar to how the sortition pool returns the group selection result. This is the natural and most common way of dealing with group members while the operator set is specific only to this case. The slice also gives some additional information that is not present in the set, like the operator's seat count. Although changing it to a set can seem justified at the moment, keeping it as a slice gives us more flexibility for the future in case we need to refactor this code. At the same time, the computational efficiency difference is rather negligible for such a small number of items (100) so it may be not worth touching it at all.